docs/developer/best-practices/navigation.asciidoc

```diff
@@ -0,0 +1,205 @@
+[[kibana-navigation]]
+== Routing, Navigation and URL
+
+{kib} platform provides a set of tools to help developers built consistent experience around routing and browser navigation.
+Some of that tooling is inside `core`, some is available as part of various plugins.  
+
+The purpose of this guide is to give you (a {kib} contributor) a high-level overview of available tools and to explain common approaches for handling routing,
+and browser navigation.
+
+This guide covers following topics:
+
+* <<navigating-between-kibana-apps>>
+* <<deep-linking>>
+* <<routing>>
+* <<history-and-location>>
+* <<state-sync>>
+* <<preserve-state>>
+
+
+[[navigating-between-kibana-apps]]
+=== Navigating between {kib} apps
+
+{kib} is a single page application and there is a set of simple rules developers should follow
+to make sure there is no page reload when navigating from one place in {kib} to another. 
+
+For example, navigation using native browser apis:  
+
+[source,js]
+----
+window.location.href = core.http.basePath.prepend(`/dashboard/my-dashboard`); // (try to avoid this)
+----
+
+would cause a full page reload.
+
+To navigate between different {kib} apps without a page reload there are apis in `core`:
+
+* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetoapp.md[core.application.navigateToApp]
+* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md[core.application.navigateToUrl]
+
+*Rendering a link to a different {kib} app on its own would also cause a full page reload:*
+
+[source,typescript jsx]
+----
+const myLink = () => 
+  <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>; 
+----
+
+A workaround could be to handle a click, prevent browser navigation and use `core.application.navigateToApp` api:
+
+[source,typescript jsx]
+----
+const MySPALink = () => 
+  <a 
+    href={core.http.basePath.prepend(`/dashboard/my-dashboard`)} 
+    onClick={(e) => {
+      e.preventDefault();
+      core.application.navigateToApp('dashboard', {path: '/my-dashboard'}); 
+    }}
+  > 
+    Go to Dashboard 
+  </a>;
+----
+
+As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy wrapper that helps with it: 
+{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].
```

Comment:
Awesome! 

⏳  Would be nice to have API docs for this plugin, or READMEs, so we aren't linking to code, but awesome that we have this here. Just thinking from someone who is starting in github, or exploring the kibana-react plugin, nothing mentioned about this service inside https://github.com/elastic/kibana/blob/master/src/plugins/kibana_react/README.md.

Reply:
At least adding `RedirectAppLinks` to https://github.com/elastic/kibana/blob/master/src/plugins/kibana_react/README.md